src/core/pattern.js

```diff
@@ -186,13 +185,62 @@ class RadialAxialShading extends BaseShading {
     const color = new Float32Array(cs.numComps),
       ratio = new Float32Array(1);
     let rgbColor;
-    for (let i = 0; i <= NUMBER_OF_SAMPLES; i++) {
+
+    let iBase = 0;
+    ratio[0] = t0;
+    fn(ratio, 0, color, 0);
+    let rgbColorBase = cs.getRgb(color, 0);
+    const cssColorBase = Util.makeHexColor(
+      rgbColorBase[0],
+      rgbColorBase[1],
+      rgbColorBase[2]
+    );
+    colorStops.push([0, cssColorBase]);
+
+    let iPrev = 1;
+    ratio[0] = t0 + step;
+    fn(ratio, 0, color, 0);
+    let rgbColorPrev = cs.getRgb(color, 0);
+
+    for (let i = 2; i < NUMBER_OF_SAMPLES; i++) {
       ratio[0] = t0 + i * step;
       fn(ratio, 0, color, 0);
       rgbColor = cs.getRgb(color, 0);
-      const cssColor = Util.makeHexColor(rgbColor[0], rgbColor[1], rgbColor[2]);
-      colorStops.push([i / NUMBER_OF_SAMPLES, cssColor]);
+
+      // See if rgbColorPrev is predicted by lerping rgbColorBase and rgbColor.
+      // If all components can be interpolated within 1, drop it.
+      // Otherwise add it as a color stop and make it the new base.
+      const scale = (iPrev - iBase) / (i - iBase);
+      const base = rgbColorBase;
+      const prev = rgbColorPrev;
+      const rgbColorPrevCanBeInterpolated =
+        Math.abs(base[0] + (rgbColor[0] - base[0]) * scale - prev[0]) <= 1 &&
+        Math.abs(base[1] + (rgbColor[1] - base[1]) * scale - prev[1]) <= 1 &&
+        Math.abs(base[2] + (rgbColor[2] - base[2]) * scale - prev[2]) <= 1;
+
+      if (!rgbColorPrevCanBeInterpolated) {
+        const cssColor = Util.makeHexColor(
+          rgbColorPrev[0],
+          rgbColorPrev[1],
+          rgbColorPrev[2]
+        );
+        colorStops.push([iPrev / NUMBER_OF_SAMPLES, cssColor]);
+
+        // TODO: When fn fequency is high (iPrev - iBase === 1 twice in a row),
```

Comment:
I'm guessing you meant "frequency" here?

Reply:
Indeed, need to clean my keyboard.